### PR TITLE
*: restructure API

### PR DESCRIPTION
**Pull request by dominikh**
This change completely overhauls the API.

It gets rid of the server and client distinction. From a traditional
networking standpoint, there are no ARP clients and servers. Each device
will send requests, and it will answer to requests. From a Go
perspective, having the distinction made some sense – many users will
probably only ever send requests, not answer them. However, the
distinction made it harder for people who do have to do both. It also
lacked a way for sending many requests at once with a single client.
While this could've been fixed by adding more APIs, it seemed a lot more
straightforward to unify the client and server responsibilities. This
gets rid of the server, the muxer and the request type.

Instead, we provide the 2 basic building blocks (reading packets and
writing packets) and some high level functions:

- Read, which reads a single packet. The packet may be a request, or it
  may be a response. The user can determine this based on the Operation
  and act accordingly. This allows to work as a server, or as a client
  dealing with many responses, or a mixture of both.

- Write, which will send a packet.

- Request, which sends a single request. It does not wait for a
  response. The response will be read via Read.

- Respond, which will send a response to a request.

- Resolve, which will send a single request, then wait for a
  response. This provides a high-level way for resolving an IP to a MAC
  address, without having to deal with individual packets or filtering
  out other noise. On the other hand, it cannot be used concurrently
  with Read, as the two would conflict with each other. It's an
  either-or choice.

Additionally, the following helpers were added:

- IsReply, to check if a packet is a response to a request.
- IsReplyFrom, like IsReply but also check for a specific IP.

### CODE REVIEW

**Comment by dominikh** [comment]
Please do not merge yet. While I'm happy with the API, I have yet to update `cmd/...`, and the purpose of this PR is to gather feedback on the API first.

File: client.go [hunk 1]
```diff
@@ -66,85 +66,130 @@ func (c *Client) Close() error {
...
+}
+
+// Respond constructs and sends a response to an ARP request. On the
+// ARP layer, it will be addressed to the sender address of the
+// packet. On the ethernet layer, it will be sent to the actual remote
+// address from which the request was received.
+//
+// For more fine-grained control, use WriteTo to write a custom
+// response.
+func (c *Client) Respond(req *Packet, hwAddr net.HardwareAddr, ip net.IP) error {
```
Comment (dominikh) [inline client.go]: Do we want to call this method `Respond` or `Reply`? When dealing with inbound packets, we call them replies (c.f. `IsReply`) – we should probably choose one and stick with it.

File: cmd/proxyarpd/main.go [hunk 2]
```diff
@@ -31,47 +32,44 @@ func main() {
...
-			return
+		if pkt.Operation != arp.OperationRequest {
+			continue
 		}
 
 		// Ignore ARP requests which are not broadcast or bound directly for
 		// this machine
-		if !bytes.Equal(r.TargetHardwareAddr, ethernet.Broadcast) && !bytes.Equal(r.TargetHardwareAddr, ifi.HardwareAddr) {
-			return
+		if !bytes.Equal(pkt.TargetHardwareAddr, ethernet.Broadcast) && !bytes.Equal(pkt.TargetHardwareAddr, ifi.HardwareAddr) {
```
Comment (dominikh) [inline cmd/proxyarpd/main.go]: Do note that this is wrong. The ethernet destination should be the broadcast address, yes. but the target hardware address should be all zeroes. The second check is likely to be wrong, too, and should also refer to the ethernet header instead.

Will fix this in a separate PR, though, to keep this one clean from unrelated changes.

File: cmd/proxyarpd/main.go [hunk 3]
```diff
@@ -31,47 +32,44 @@ func main() {
...
-		p, err := arp.NewPacket(
-			arp.OperationReply,
-			ifi.HardwareAddr,
-			ip,
-			r.SenderHardwareAddr,
-			r.SenderIP,
-		)
-		if err != nil {
-			log.Fatal(err)
+		if !bytes.Equal(pkt.TargetIP, ip) {
```
Comment (dominikh) [inline cmd/proxyarpd/main.go]: Will change to IP.Equal in a separate PR.

File: client.go [hunk 4]
```diff
@@ -66,85 +66,130 @@ func (c *Client) Close() error {
...
-// of a machine using its IPv4 address.
-func (c *Client) Request(ip net.IP) (net.HardwareAddr, error) {
+// Request sends an ARP request, asking for the hardware address
+// associated with an IPv4 address. The response, if any, can be read
+// with the Read method.
+//
+// Unlike Resolve, which provides an easier interface for getting the
+// hardware address, Request allows sending many requests in a row,
+// retrieving the responses afterwards.
+func (c *Client) Request(ip net.IP) error {
```
Comment (dominikh) [inline client.go]: Request or SendRequest?

File: client.go (hunk 1 above)
Comment (dominikh) [inline client.go]: Alternatively it could also be SendResponse or SendReply (same question as for Request/SendRequest)

File: client.go (hunk 4 above)
Comment (mdlayher) [inline client.go]: Request seems fine to me.

File: client.go [hunk 5]
```diff
@@ -66,85 +66,130 @@ func (c *Client) Close() error {
...
+// hardware address.
+func (c *Client) IsReply(p *Packet) bool {
+	return p.Operation == OperationReply &&
+		p.TargetIP.Equal(c.ip) &&
+		bytes.Equal(p.TargetHardwareAddr, c.ifi.HardwareAddr)
+}
+
+// IsReplyFrom reports whether the packet p is a response to our query
+// for ip. It works like IsReply, but additionally checks if the
+// response is for a specific IPv4 address ip.
```
Comment (mdlayher) [inline client.go]: Should "for" be "from" here?  The code seems to indicate it's looking at `p.SenderIP`.

File: client.go (hunk 1 above)
Comment (mdlayher) [inline client.go]: I was thinking about the same thing, and I think `Reply` would be just fine.

File: cmd/proxyarpd/main.go (hunk 2 above)
Comment (mdlayher) [inline cmd/proxyarpd/main.go]: Sounds good to me.

File: packet.go [hunk 6]
```diff
@@ -63,6 +69,10 @@ type Packet struct {
 
 	// TargetIP specifies the IPv4 address of the target of this Packet.
 	TargetIP net.IP
+
+	// RemoteAddr specifies the actual remote address from which the
+	// ARP packet was received.
```
Comment (mdlayher) [inline packet.go]: Can we add another note here that says something along the lines of: "RemoteAddr does not appear in the binary form of a Packet".

File: client.go (hunk 5 above)
Comment (dominikh) [inline client.go]: For and From are kind of synonymous here, aren't they? If we ask *for* an IP and get a response, that response will come *from* the system owning that IP. 

But I suppose For is less ambiguous, with ARP spoofing and all.

**Comment by mdlayher** [comment]
I'm happy to merge this once a comment is added here: https://github.com/mdlayher/arp/pull/2#discussion_r36769471

File: client.go (hunk 5 above)
Comment (dominikh) [inline client.go]: Oh… I read you backwards. You're confused by the "to our query for ip" in the docstring? We send a query, asking for the owner of an IP. So p is a response to our query for ip. But I see the problem, we are using for and from interchangeably here. I'll rename the method to IsReplyFor, then the docstring should make more sense overall.

**Comment by dominikh** [comment]
I pushed new commits addressing your comments. If you're fine with them, let me know and I'll squash them into the original commit, before you merge.
